pkg/apply/filter/filter.go

```diff
@@ -16,4 +16,6 @@ type ValidationFilter interface {
 	Name() string
 	// Filter returns true if validation fails or an error.
 	Filter(obj *unstructured.Unstructured) (bool, error)
+	// Reason returns the string explanation if filtered.
+	Reason() string
```

Comment:
So I find this a bit strange. We are using a single instance of a filter to handle all resources, but this adds state specific to a single resource. This seems like a solution that can easily lead to bugs.

Reply:
You're right. It is not re-entrant. I will change the signature of the `Filter()` function to additionally return the reason string.